jsonstar/JsonStar.Schema.Ast.ToJsonSchema.fst

```diff
@@ -0,0 +1,197 @@
+    // NOTE: The rest of refinement types is not applicable to strings so we ignore them. 
+    //       Ideally, we would make them unrepresentable, but I don't know how to define 
+    //       AST.typ in a way that allows that as is still practical to use. 
+    | _ -> opt
+
+let rec refineNumber (x : AST.refinement_type) (opt : number_options) : Tot number_options = 
+    match x with
+    | AST.NumberRefinement r -> begin
+        match r with
+            | AST.Maximum v -> { opt with maximum = min_opt opt.maximum (Some (Int v)) }
+            | AST.Minimum v -> { opt with minimum = max_opt opt.minimum (Some (Int v)) }
+        end
+    | AST.And lr rr -> refineNumber lr (refineNumber rr opt)
+    | _ -> opt
+
+let rec refineEnum (x : AST.refinement_type) (vs : list string) : Tot (list string) = 
+    match x with
+    | AST.EnumRefinement sr -> begin
+        match sr with
+            | AST.Allow values -> L.filter (fun x -> L.mem x values) vs
+            | AST.Disallow values -> L.filter (fun x -> not (L.mem x values)) vs
+        end
+    | AST.And lr rr -> refineEnum lr (refineEnum rr vs)
+    | _ -> vs
+
+let primitiveToJsonSchema (x : AST.primitive) : Tot schema = 
+    let _typ = 
+        match x with
+        | AST.Int -> Number (Mknumber_options None None)
+        | AST.String -> String (mkempty_string_options ())
+        | AST.Boolean -> Boolean
+    in
+    {
+        _id         = None;
+        _schema     = None;
+        _type       = _typ;
+        description = None;
+        title       = None;
+        _default    = None;
+        readonly    = None;
+        format      = None;
+        definitions = [];
+    }
+
+let enumToJsonSchema (x : AST.enum) : Tot schema = 
+    // TODO: Translate attributes into schema properties
+    let values = (let open AST in x.enum_values) in
+    {
+        _id         = None;
+        _schema     = None;
+        _type       = Enum values;
+        description = None;
+        title       = None;
+        _default    = None;
+        readonly    = None;
+        format      = None;
+        definitions = [];
+    }
+
+let rec refinementToJsonSchema (x : AST.refinement) : Tot schema = 
+    let baseTyp = (let open AST in x._base) in
+    let baseSchema = toJsonSchema baseTyp in
+    let ref = (let open AST in x._refinement) in
+    // NOTE: Ideally we'd change 'AST.refinement' to only allow certain refinement types
+    //       depending on the refined type. Unfortunately this is not possible with F* 
+    //       at least currently. 
+    match baseSchema._type with
+    | String opt -> { baseSchema with _type = String (refineString ref opt); }
+    | Integer -> baseSchema
+    | Number opt -> { baseSchema with _type = Number (refineNumber ref opt); }
+    | Enum vs -> { baseSchema with _type = Enum (refineEnum ref vs); }
+    | Boolean -> baseSchema
+    // TODO: Apply other refinements
+    | _ -> baseSchema
+    
+and typedefToJsonSchema (x : AST.typedef) : Tot schema = 
```

Comment:
After you factor some of these types out of the big mutual inductive block (as I suggested on Friday), don't forget to do the same for some of these functions.